#5 quality gate

lowellfarrell opened the pull request:
A functional overall in usage key changes:
- Adding of quality gate feature
- pylint-exit-options wraps around pylint and 
- pylint option passed to pylint-exit-options will be forwarded to pylint
- You no longer need to use '||' to pipe pylint output to pylint-exit-options

lowellfarrell committed:
Update pylint_exit_options.py

Feature add: Quality Gate: allows users to set a minimum code quality threshold Refactor the the utility is called make it one command deprecate the use of '||'

lowellfarrell committed:
Update README.md

Updating Read Me to reflect added functionality and latest refactors

lowellfarrell committed:
Update pylint_exit_options.py

Reworking cli output

lowellfarrell committed:
Update requirements.txt

Updating requirements - adding pylint

lowellfarrell committed:
Update setup.py

Updating required packages - adding pylint

lowellfarrell committed:
Update Jenkinsfile

Updating Jenkins CI pipeline to change 'pylint-exit-options' usage the reflected new usage pattern from latest refactor.

ActionCactus commented on pylint_exit_options.py:
```diff
@@ -1,269 +1,302 @@
+
+        if exit_code:
+            print('The following types of issues are blocking:')
+            print('')
```
Could probably condense these two statements

lowellfarrell replied:
Thanks, i'll change this

lowellfarrell replied:
refactored this to to this """print('The following types of issues are blocking:' + os.linesep)"""

ActionCactus commented on pylint_exit_options.py:
```diff
@@ -1,269 +1,302 @@
+            arg_values (List): A list of setting passed into the '--exit-report' option which can change which exit
+            codes will be returned to cli
+        """
+        if 'F' in arg_values:
```
Why not use a `Letter => {PassSeverity, FailSeverity}` dictionary and loop through them?

lowellfarrell replied:
Thanks, I'll change this

lowellfarrell replied:
I reworked this the code looks like this now

"""
        check = {
            'F': __FATAL__,
            'E': __ERROR__,
            'W': __WARNING__,
            'R': __REFACTOR__,
            'C': __CONVENTION__,
            'U': __USAGE__
        }
        for key, value in check.items():
            if key in arg_values:
                self._apply_enforcement_setting(value, value)
            else:
                self._apply_enforcement_setting(value, __SUPPRESS__)
"""

ActionCactus commented on pylint_exit_options.py:
```diff
@@ -1,269 +1,302 @@
+            value (int): new value for level
+
+        """
+        positions = {
```
Perhaps this could be a class variable so that you don't re-initialize it every time the method is called?  Not sure if Python handles these sorts of things differently, though.

lowellfarrell replied:
The position array is completely gone now

ActionCactus commented on pylint_exit_options.py:
```diff
@@ -1,269 +1,302 @@
+        position = positions[key]
+
+        # unpack the tuple so it can be modified
+        encoded, description, enforce = self.exit_value_defaults[position]
```
Nice use of tuple returns

lowellfarrell committed:
Update pylint_exit_options.py

-adding type hinting
-fixing output sting for quality gate failures

lowellfarrell committed:
Update pylint_exit_options.py

refactoring _set_report_arg_values function to reduce if else blocks

lowellfarrell committed:
Update pylint_exit_options.py

A slight refactor of the exit_value_defaults and how we use it

lowellfarrell committed:
Update pylint_exit_options.py

fixing TypeError: 'dict' object is not callable:

Traceback (most recent call last):
  File "citools/pylintexitoptions.py", line 283, in <module>
    main()
  File "citools/pylintexitoptions.py", line 273, in main
    exit_code = ex.handle_exit_code(run.linter.msg_status) + quality.check_quality(run.linter.stats['global_note'])
  File "citools/pylintexitoptions.py", line 141, in handle_exit_code
    messages = self._get_messages(value)
  File "citools/pylintexitoptions.py", line 85, in _get_messages
    return [x[1] for x in self._decode(value)]
  File "citools/pylintexitoptions.py", line 65, in _decode
    return [self.exit_value_defaults(x[1]) for x in zip(bitarray(bin(value)[2:])[::-1], self.exit_value_defaults) if
  File "citools/pylintexitoptions.py", line 66, in <listcomp>
    x[0]]

lowellfarrell committed:
[#5] adding some type hinting and reworking some 'print('')' statements

lowellfarrell committed:
Cleaning the setup.py file to minimize the requirements for running it
